sql: panic in `RIGHT JOIN`

Found a simple panic in `RIGHT JOIN` while investigating #12311.

Setup:
```
create table a (i int);
create table b (i int, b bool);
insert into a values (1), (2), (3);
insert into b values (1, false), (2, false), (3, true);
```

Panic:
```
root@:26257> SELECT a.i, b.b FROM a RIGHT JOIN b ON a.i=b.i;
panic: invalid setNeededColumns argument (len 2 instead of 3): [true false] [recovered]
        panic: SELECT a.i, b.i FROM a RIGHT JOIN b ON a.i = b.i;: invalid setNeededColumns argument (len 2 instead of 3): [true false] [recovered]

goroutine 184 [running]:
panic(0x527cba0, 0xc420200e00)
        /usr/local/Cellar/go/1.7.4/libexec/src/runtime/panic.go:500 +0x1a1
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc4202ae000)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:185 +0x6e
panic(0x527cba0, 0xc420200e00)
        /usr/local/Cellar/go/1.7.4/libexec/src/runtime/panic.go:458 +0x243
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).ExecuteStatements.func1(0xc4207fb0da, 0x31)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:446 +0x149
panic(0x5236600, 0xc420200de0)
        /usr/local/Cellar/go/1.7.4/libexec/src/runtime/panic.go:458 +0x243
github.com/cockroachdb/cockroach/pkg/sql.(*scanNode).setNeededColumns(0xc420166700, 0xc420200db8, 0x2, 0x8)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/scan.go:311 +0x25c
github.com/cockroachdb/cockroach/pkg/sql.(*joinNode).setNeededColumns(0xc4204c5a40, 0xc420200db0, 0x5, 0x5)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/join.go:273 +0xb1
github.com/cockroachdb/cockroach/pkg/sql.(*selectNode).expandPlan(0xc4209c6b00, 0x0, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/select.go:412 +0x196
github.com/cockroachdb/cockroach/pkg/sql.(*selectTopNode).expandPlan(0xc420bcd2c0, 0x0, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/select_top.go:77 +0x61
github.com/cockroachdb/cockroach/pkg/sql.(*planner).makePlan(0xc4207ce170, 0x5d5b0a0, 0xc420c1d3e0, 0x28f5b901, 0x6241ae0, 0xc420bddf38, 0x4343102, 0xecfe26809)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:232 +0xd0
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmt(0xc420276840, 0x5d5b0a0, 0xc420c1d3e0, 0xc4207ce170, 0xc420bde101, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1283 +0xa3
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmtInOpenTxn(0xc420276840, 0x5d5b0a0, 0xc420c1d3e0, 0xc4207ce170, 0x101, 0xc4207ce068, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1090 +0x2c2
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmtsInCurrentTxn(0xc420276840, 0xc420200bc0, 0x1, 0x1, 0xc4207ce170, 0xc4207ce068, 0x101, 0x6, 0xc420bdebf8, 0x404ad2a, ...)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:836 +0x7cd
github.com/cockroachdb/cockroach/pkg/sql.runTxnAttempt(0xc420276840, 0xc4207ce170, 0x1, 0xc4207ce068, 0xc420200c00, 0xc420200bc0, 0x1, 0x1, 0x0, 0x0, ...)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:755 +0xeb
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execRequest.func2(0xc4209c69a0, 0xc420200c00, 0x4000000000000000, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:609 +0x22f
github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Exec(0xc4209c69a0, 0xc420c50101, 0xc4204a0a00, 0xc420c520c0, 0x0, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:520 +0x215
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execRequest(0xc420276840, 0xc4207ce000, 0xc4207fb0da, 0x31, 0x0, 0x0, 0x0, 0x0, 0xc420a69400)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:631 +0x572
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).ExecuteStatements(0xc420276840, 0xc4207ce000, 0xc4207fb0da, 0x31, 0x0, 0x0, 0x0, 0x0, 0xc4207fb000)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:452 +0x11f
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).executeStatements(0xc4201f2400, 0x954d5c0, 0xc420a04030, 0xc4207fb0da, 0x31, 0x0, 0x0, 0x0, 0x0, 0x4658301, ...)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:746 +0x94
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).handleSimpleQuery(0xc4201f2400, 0x954d5c0, 0xc420a04030, 0xc4201f2428, 0x36, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:420 +0xb3
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).serve(0xc4201f2400, 0x954d5c0,
<snip>
```

--- a/pkg/sql/join.go
+++ b/pkg/sql/join.go
@@ -355,11 +355,13 @@ func (n *joinNode) Start() error {
 
 	// If needed, pre-allocate left and right rows of NULL tuples for when the
 	// join predicate fails to match.
-	if n.joinType != joinTypeInner {
+	if n.joinType == joinTypeLeftOuter || n.joinType == joinTypeFullOuter {
 		n.emptyRight = make(parser.DTuple, len(n.right.plan.Columns()))
 		for i := range n.emptyRight {
 			n.emptyRight[i] = parser.DNull
 		}
+	}
+	if n.joinType == joinTypeRightOuter || n.joinType == joinTypeFullOuter {
 		n.emptyLeft = make(parser.DTuple, len(n.left.plan.Columns()))
 		for i := range n.emptyLeft {
 			n.emptyLeft[i] = parser.DNull

--- a/pkg/sql/join.go
+++ b/pkg/sql/join.go
@@ -36,22 +36,22 @@ const (
 
 // bucket here is the set of rows for a given group key (comprised of
 // columns specified by the join constraints), 'seen' is used to determine if
-// there was a matching group (with the same group key) in the opposite stream.
+// there was a matching row in the opposite stream.
 type bucket struct {
 	rows []parser.DTuple
-	seen bool
+	seen []bool
 }
 
-func (b *bucket) Seen() bool {
-	return b.seen
+func (b *bucket) Seen(i int) bool {
+	return b.seen[i]
 }
 
 func (b *bucket) Rows() []parser.DTuple {
 	return b.rows
 }
 
-func (b *bucket) MarkSeen() {
-	b.seen = true
+func (b *bucket) MarkSeen(i int) {
+	b.seen[i] = true
 }
 
 func (b *bucket) AddRow(row parser.DTuple) {
@@ -86,6 +86,14 @@ func (b *buckets) AddRow(acc WrappedMemoryAccount, encoding []byte, row parser.D
 		b.buckets[string(encoding)] = bk
 	}
 	return nil
+}
+
+// InitSeen must be run before the buckets are used to initialize the rows'
+// seen arrays.
+func (b *buckets) InitSeen() {
+	for _, bucket := range b.buckets {
+		bucket.seen = make([]bool, len(bucket.rows))
+	}
 }
 
 func (b *buckets) Close() {
@@ -371,7 +379,7 @@ func (n *joinNode) hashJoinStart() error {
 			return err
 		}
 		if !hasRow {
-			return nil
+			break
 		}
 		row := n.right.plan.Values()
 		encoding, _, err := n.pred.encode(scratch, row, n.pred.rightEqualityIndices)
@@ -385,6 +393,8 @@ func (n *joinNode) hashJoinStart() error {
 
 		scratch = encoding[:0]
 	}
+	n.buckets.InitSeen()
+	return nil
 }
 
 func (n *joinNode) debugNext() (bool, error) {
@@ -512,12 +522,11 @@ func (n *joinNode) Next() (res bool, err error) {
 			}
 			return n.buffer.Next(), nil
 		}
-		b.MarkSeen()
 
 		// We iterate through all the rows in the bucket attempting to match the
 		// filter, if the filter passes we add it to the buffer.
 		foundMatch := false
-		for _, rrow := range b.Rows() {
+		for idx, rrow := range b.Rows() {
 			passesFilter, err := n.pred.eval(&n.planner.evalCtx, n.output, lrow, rrow)
 			if err != nil {
 				return false, err
@@ -529,6 +538,10 @@ func (n *joinNode) Next() (res bool, err error) {
 			foundMatch = true
 
 			n.pred.prepareRow(n.output, lrow, rrow)
+			// TODO(jordan) MarkSeen is not required unless the join is RIGHT
+			// or FULL. In these cases, we can skip allocating the seen arrays
+			// for every bucket.
+			b.MarkSeen(idx)
 			if _, err := n.buffer.AddRow(n.output); err != nil {
 				return false, err
 			}
@@ -554,14 +567,14 @@ func (n *joinNode) Next() (res bool, err error) {
 	}
 
 	for _, b := range n.buckets.Buckets() {
-		if !b.Seen() {
-			for _, rrow := range b.Rows() {
+		for idx, rrow := range b.Rows() {
+			if !b.Seen(idx) {
 				n.pred.prepareRow(n.output, n.emptyLeft, rrow)
 				if _, err := n.buffer.AddRow(n.output); err != nil {
 					return false, err
 				}
+				b.MarkSeen(idx)
 			}
-			b.MarkSeen()
 		}
 	}
 

--- a/pkg/sql/join.go
+++ b/pkg/sql/join.go
@@ -30,6 +30,7 @@ type joinType int
 const (
 	joinTypeInner joinType = iota
 	joinTypeLeftOuter
+	joinTypeRightOuter
 	joinTypeFullOuter
 )
 
@@ -105,9 +106,8 @@ type joinNode struct {
 	joinType joinType
 
 	// The data sources.
-	left    planDataSource
-	right   planDataSource
-	swapped bool
+	left  planDataSource
+	right planDataSource
 
 	// pred represents the join predicate.
 	pred *joinPredicate
@@ -170,9 +170,6 @@ func commonColumns(left, right *dataSourceInfo) parser.NameList {
 func (p *planner) makeJoin(
 	astJoinType string, left planDataSource, right planDataSource, cond parser.JoinCond,
 ) (planDataSource, error) {
-	leftInfo, rightInfo := left.info, right.info
-
-	swapped := false
 	var typ joinType
 	switch astJoinType {
 	case "JOIN", "INNER JOIN", "CROSS JOIN":
@@ -180,18 +177,18 @@ func (p *planner) makeJoin(
 	case "LEFT JOIN":
 		typ = joinTypeLeftOuter
 	case "RIGHT JOIN":
-		left, right = right, left // swap
-		typ = joinTypeLeftOuter
-		swapped = true
+		typ = joinTypeRightOuter
 	case "FULL JOIN":
 		typ = joinTypeFullOuter
 	default:
 		return planDataSource{}, errors.Errorf("unsupported JOIN type %T", astJoinType)
 	}
 
+	leftInfo, rightInfo := left.info, right.info
+
 	// Check that the same table name is not used on both sides.
-	for _, alias := range right.info.sourceAliases {
-		if _, ok := left.info.sourceAliases.srcIdx(alias.name); ok {
+	for _, alias := range rightInfo.sourceAliases {
+		if _, ok := leftInfo.sourceAliases.srcIdx(alias.name); ok {
 			t := alias.name.Table()
 			if t == "" {
 				// Allow joins of sources that define columns with no
@@ -235,7 +232,6 @@ func (p *planner) makeJoin(
 		joinType: typ,
 		pred:     pred,
 		columns:  info.sourceColumns,
-		swapped:  swapped,
 	}
 
 	n.buffer = &RowBuffer{
@@ -296,11 +292,9 @@ func (n *joinNode) ExplainPlan(v bool) (name, description string, children []pla
 		}
 		buf.WriteString(jType)
 	case joinTypeLeftOuter:
-		if !n.swapped {
-			buf.WriteString("LEFT OUTER")
-		} else {
-			buf.WriteString("RIGHT OUTER")
-		}
+		buf.WriteString("LEFT OUTER")
+	case joinTypeRightOuter:
+		buf.WriteString("RIGHT OUTER")
 	case joinTypeFullOuter:
 		buf.WriteString("FULL OUTER")
 	}
@@ -308,9 +302,6 @@ func (n *joinNode) ExplainPlan(v bool) (name, description string, children []pla
 	n.pred.format(&buf)
 
 	subplans := []planNode{n.left.plan, n.right.plan}
-	if n.swapped {
-		subplans[0], subplans[1] = subplans[1], subplans[0]
-	}
 	subplans = n.planner.collectSubqueryPlans(n.pred.filter, subplans)
 
 	return "join", buf.String(), subplans
@@ -356,7 +347,7 @@ func (n *joinNode) Start() error {
 
 	// If needed, pre-allocate left and right rows of NULL tuples for when the
 	// join predicate fails to match.
-	if n.joinType == joinTypeLeftOuter || n.joinType == joinTypeFullOuter {
+	if n.joinType != joinTypeInner {
 		n.emptyRight = make(parser.DTuple, len(n.right.plan.Columns()))
 		for i := range n.emptyRight {
 			n.emptyRight[i] = parser.DNull
@@ -491,7 +482,7 @@ func (n *joinNode) Next() (res bool, err error) {
 		//    | NULL |  52  |
 		//    | NULL |  52  |
 		if containsNull {
-			if n.joinType == joinTypeInner {
+			if n.joinType == joinTypeInner || n.joinType == joinTypeRightOuter {
 				scratch = encoding[:0]
 				// Failed to match -- no matching row, nothing to do.
 				continue
@@ -507,11 +498,11 @@ func (n *joinNode) Next() (res bool, err error) {
 
 		b, ok := n.buckets.Fetch(encoding)
 		if !ok {
-			if n.joinType == joinTypeInner {
+			if n.joinType == joinTypeInner || n.joinType == joinTypeRightOuter {
 				scratch = encoding[:0]
 				continue
 			}
-			// Outer join: unmatched rows are padded with NULLs.
+			// Left or full outer join: unmatched rows are padded with NULLs.
 			// Given that we did not find a matching right row we append an
 			// empty right row to the left row, adding the result to our buffer
 			// for the subsequent call to Next().
@@ -542,9 +533,10 @@ func (n *joinNode) Next() (res bool, err error) {
 				return false, err
 			}
 		}
-		if !foundMatch && n.joinType != joinTypeInner {
-			// If none of the rows matched the filter and we are computed an outer
-			// join, we need to add an row with an empty right side.
+		if !foundMatch && n.joinType != joinTypeInner && n.joinType != joinTypeRightOuter {
+			// If none of the rows matched the filter and we are computing a
+			// left or full outer join, we need to add a row with an empty
+			// right side.
 			n.pred.prepareRow(n.output, lrow, n.emptyRight)
 			if _, err := n.buffer.AddRow(n.output); err != nil {
 				return false, err
@@ -557,7 +549,7 @@ func (n *joinNode) Next() (res bool, err error) {
 	}
 
 	// no more lrows, we go through the unmatched rows in the internal hashmap.
-	if n.joinType != joinTypeFullOuter {
+	if n.joinType == joinTypeInner || n.joinType == joinTypeLeftOuter {
 		return false, nil
 	}
 

--- a/pkg/sql/join.go
+++ b/pkg/sql/join.go
@@ -21,6 +21,8 @@ import (
 	"bytes"
 	"fmt"
 
+	"unsafe"
+
 	"github.com/cockroachdb/cockroach/pkg/sql/parser"
 	"github.com/pkg/errors"
 )
@@ -88,12 +90,19 @@ func (b *buckets) AddRow(acc WrappedMemoryAccount, encoding []byte, row parser.D
 	return nil
 }
 
+const sizeOfBoolSlice = unsafe.Sizeof([]bool{})
+const sizeOfBool = unsafe.Sizeof(true)
+
 // InitSeen initializes the seen array for each of the buckets. It must be run
 // before the buckets' seen state is used.
-func (b *buckets) InitSeen() {
+func (b *buckets) InitSeen(acc WrappedMemoryAccount) error {
 	for _, bucket := range b.buckets {
+		if err := acc.Grow(int64(sizeOfBoolSlice + uintptr(len(bucket.rows))*sizeOfBool)); err != nil {
+			return err
+		}
 		bucket.seen = make([]bool, len(bucket.rows))
 	}
+	return nil
 }
 
 func (b *buckets) Close() {
@@ -400,7 +409,7 @@ func (n *joinNode) hashJoinStart() error {
 		scratch = encoding[:0]
 	}
 	if n.joinType == joinTypeFullOuter || n.joinType == joinTypeRightOuter {
-		n.buckets.InitSeen()
+		return n.buckets.InitSeen(acc)
 	}
 	return nil
 }
@@ -590,7 +599,6 @@ func (n *joinNode) Next() (res bool, err error) {
 				if _, err := n.buffer.AddRow(n.output); err != nil {
 					return false, err
 				}
-				b.MarkSeen(idx)
 			}
 		}
 	}

--- a/pkg/sql/join.go
+++ b/pkg/sql/join.go
@@ -88,8 +88,8 @@ func (b *buckets) AddRow(acc WrappedMemoryAccount, encoding []byte, row parser.D
 	return nil
 }
 
-// InitSeen must be run before the buckets are used to initialize the rows'
-// seen arrays.
+// InitSeen initializes the seen array for each of the buckets. It must be run
+// before the buckets' seen state is used.
 func (b *buckets) InitSeen() {
 	for _, bucket := range b.buckets {
 		bucket.seen = make([]bool, len(bucket.rows))
@@ -393,7 +393,9 @@ func (n *joinNode) hashJoinStart() error {
 
 		scratch = encoding[:0]
 	}
-	n.buckets.InitSeen()
+	if n.joinType == joinTypeFullOuter || n.joinType == joinTypeRightOuter {
+		n.buckets.InitSeen()
+	}
 	return nil
 }
 
@@ -541,10 +543,11 @@ func (n *joinNode) Next() (res bool, err error) {
 			foundMatch = true
 
 			n.pred.prepareRow(n.output, lrow, rrow)
-			// TODO(jordan) MarkSeen is not required unless the join is RIGHT
-			// or FULL. In these cases, we can skip allocating the seen arrays
-			// for every bucket.
-			b.MarkSeen(idx)
+			if wantUnmatchedRight {
+				// Mark the row as seen if we need to retrieve the rows
+				// without matches for right or full joins later.
+				b.MarkSeen(idx)
+			}
 			if _, err := n.buffer.AddRow(n.output); err != nil {
 				return false, err
 			}

--- a/pkg/sql/join.go
+++ b/pkg/sql/join.go
@@ -149,6 +149,10 @@ type joinNode struct {
 	// doneReadingRight is used by debugNext() and DebugValues() when
 	// explain == explainDebug.
 	doneReadingRight bool
+
+	// finishedOutput indicates that we've finished writing all of the rows for
+	// this join and that we can quit as soon as our buffer is empty.
+	finishedOutput bool
 }
 
 // commonColumns returns the names of columns common on the
@@ -422,6 +426,17 @@ func (n *joinNode) Next() (res bool, err error) {
 		return n.debugNext()
 	}
 
+	// If results available from from previously computed results, we just
+	// return true.
+	if n.buffer.Next() {
+		return true, nil
+	}
+
+	// If the buffer is empty and we've finished outputting, we're done.
+	if n.finishedOutput {
+		return false, nil
+	}
+
 	wantUnmatchedLeft := n.joinType == joinTypeLeftOuter || n.joinType == joinTypeFullOuter
 	wantUnmatchedRight := n.joinType == joinTypeRightOuter || n.joinType == joinTypeFullOuter
 
@@ -430,12 +445,6 @@ func (n *joinNode) Next() (res bool, err error) {
 			// No rows on right; don't even try.
 			return false, nil
 		}
-	}
-
-	// If results available from from previously computed results, we just
-	// return true.
-	if n.buffer.Next() {
-		return true, nil
 	}
 
 	// Compute next batch of matching rows.
@@ -585,6 +594,7 @@ func (n *joinNode) Next() (res bool, err error) {
 			}
 		}
 	}
+	n.finishedOutput = true
 
 	return n.buffer.Next(), nil
 }

--- a/pkg/sql/join.go
+++ b/pkg/sql/join.go
@@ -418,8 +418,11 @@ func (n *joinNode) Next() (res bool, err error) {
 		return n.debugNext()
 	}
 
+	wantUnmatchedLeft := n.joinType == joinTypeLeftOuter || n.joinType == joinTypeFullOuter
+	wantUnmatchedRight := n.joinType == joinTypeRightOuter || n.joinType == joinTypeFullOuter
+
 	if len(n.buckets.Buckets()) == 0 {
-		if n.joinType != joinTypeLeftOuter && n.joinType != joinTypeFullOuter {
+		if !wantUnmatchedLeft {
 			// No rows on right; don't even try.
 			return false, nil
 		}
@@ -492,7 +495,7 @@ func (n *joinNode) Next() (res bool, err error) {
 		//    | NULL |  52  |
 		//    | NULL |  52  |
 		if containsNull {
-			if n.joinType == joinTypeInner || n.joinType == joinTypeRightOuter {
+			if !wantUnmatchedLeft {
 				scratch = encoding[:0]
 				// Failed to match -- no matching row, nothing to do.
 				continue
@@ -508,7 +511,7 @@ func (n *joinNode) Next() (res bool, err error) {
 
 		b, ok := n.buckets.Fetch(encoding)
 		if !ok {
-			if n.joinType == joinTypeInner || n.joinType == joinTypeRightOuter {
+			if !wantUnmatchedLeft {
 				scratch = encoding[:0]
 				continue
 			}
@@ -546,7 +549,7 @@ func (n *joinNode) Next() (res bool, err error) {
 				return false, err
 			}
 		}
-		if !foundMatch && n.joinType != joinTypeInner && n.joinType != joinTypeRightOuter {
+		if !foundMatch && wantUnmatchedLeft {
 			// If none of the rows matched the filter and we are computing a
 			// left or full outer join, we need to add a row with an empty
 			// right side.
@@ -562,7 +565,7 @@ func (n *joinNode) Next() (res bool, err error) {
 	}
 
 	// no more lrows, we go through the unmatched rows in the internal hashmap.
-	if n.joinType == joinTypeInner || n.joinType == joinTypeLeftOuter {
+	if !wantUnmatchedRight {
 		return false, nil
 	}
 

--- a/pkg/sql/testdata/join
+++ b/pkg/sql/testdata/join
@@ -90,8 +90,8 @@ SELECT * FROM onecolumn AS a(x) RIGHT OUTER JOIN onecolumn AS b(y) ON a.x = b.y
 ----
    x     y
   44    44
-NULL  NULL
   42    42
+NULL  NULL
 
 query I colnames
 SELECT * FROM onecolumn AS a RIGHT OUTER JOIN onecolumn AS b USING(x) ORDER BY x
@@ -106,8 +106,8 @@ SELECT * FROM onecolumn AS a NATURAL RIGHT OUTER JOIN onecolumn AS b
 ----
    x
   44
-NULL
   42
+NULL
 
 statement ok
 CREATE TABLE onecolumn_w(w INT); INSERT INTO onecolumn_w(w) VALUES (42),(43)
@@ -329,6 +329,43 @@ SELECT o.x, t.y FROM onecolumn o LEFT OUTER JOIN twocolumn t ON (o.x=t.x AND t.x
 44   51
 NULL NULL
 42   NULL
+
+# Simple test cases for inner, left, right, and outer joins
+
+statement ok
+CREATE TABLE a (i int); INSERT INTO a VALUES (1), (2), (3)
+
+statement ok
+CREATE TABLE b (i int, b bool); INSERT INTO b VALUES (2, true), (3, true), (4, false)
+
+query IIB
+SELECT * FROM a INNER JOIN b ON a.i = b.i
+----
+2 2 true
+3 3 true
+
+query IIB
+SELECT * FROM a LEFT OUTER JOIN b ON a.i = b.i
+----
+1 NULL NULL
+2 2    true
+3 3    true
+
+query IIB
+SELECT * FROM a RIGHT OUTER JOIN b ON a.i = b.i
+----
+2    2    true
+3    3    true
+NULL 4    false
+
+query IIB
+SELECT * FROM a FULL OUTER JOIN b ON a.i = b.i
+----
+1    NULL NULL
+2    2    true
+3    3    true
+NULL 4    false
+
 
 # Check column orders and names.
 query IIIIII colnames

--- a/pkg/sql/testdata/join
+++ b/pkg/sql/testdata/join
@@ -368,12 +368,33 @@ NULL 4    false
 
 # Full outer join with filter predicate
 query IIB
-SELECT * FROM a FULL OUTER JOIN b ON (a.i = b.i and a.i>2);
+SELECT * FROM a FULL OUTER JOIN b ON (a.i = b.i and a.i>2) ORDER BY a.i, b.i;
 ----
+NULL 2    true
+NULL 4    false
 1    NULL NULL
 2    NULL NULL
 3    3    true
-NULL 2    true
+
+# Duplicate right matches for a single left row
+statement ok
+INSERT INTO b VALUES (3, false)
+
+query IIB
+SELECT * FROM a RIGHT OUTER JOIN b ON a.i=b.i ORDER BY b.i, b.b;
+----
+2    2 true
+3    3 false
+3    3 true
+NULL 4 false
+
+query IIB
+SELECT * FROM a FULL OUTER JOIN b ON a.i=b.i ORDER BY b.i, b.b;
+----
+1    NULL NULL
+2    2    true
+3    3    false
+3    3    true
 NULL 4    false
 
 

--- a/pkg/sql/testdata/join
+++ b/pkg/sql/testdata/join
@@ -330,7 +330,7 @@ SELECT o.x, t.y FROM onecolumn o LEFT OUTER JOIN twocolumn t ON (o.x=t.x AND t.x
 NULL NULL
 42   NULL
 
-# Simple test cases for inner, left, right, and outer joins
+## Simple test cases for inner, left, right, and outer joins
 
 statement ok
 CREATE TABLE a (i int); INSERT INTO a VALUES (1), (2), (3)
@@ -364,6 +364,16 @@ SELECT * FROM a FULL OUTER JOIN b ON a.i = b.i
 1    NULL NULL
 2    2    true
 3    3    true
+NULL 4    false
+
+# Full outer join with filter predicate
+query IIB
+SELECT * FROM a FULL OUTER JOIN b ON (a.i = b.i and a.i>2);
+----
+1    NULL NULL
+2    NULL NULL
+3    3    true
+NULL 2    true
 NULL 4    false
 
 

